File: include/qmckl_gpu_f.f90
module qmckl_gpu
    use, intrinsic :: iso_c_binding
    integer, parameter :: qmckl_context_device = c_int64_t
    integer, parameter :: qmckl_exit_code = c_int32_t
    interface

    !!!!!!!!!!!
        ! CONTEXT
    !!!!!!!!!!!

        integer(qmckl_context_device) function qmckl_context_touch_device(context) &
            bind(C, name="qmckl_context_touch_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), value :: context
        end function qmckl_context_touch_device

        integer(qmckl_context_device) function qmckl_context_create_device(device_id) &
            bind(C, name="qmckl_context_create_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(c_int32_t), value :: device_id
        end function qmckl_context_create_device

        integer(qmckl_exit_code) function qmckl_context_destroy_device(context) &
            bind(C, name="qmckl_context_destroy_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), value :: context
        end function qmckl_context_destroy_device

    !!!!!!!!!!!
        ! MEMORY
    !!!!!!!!!!!

        type(c_ptr) function qmckl_malloc_device(context, size) &
            bind(C, name="qmckl_malloc_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), value :: context
            integer(c_size_t), value :: size
        end function qmckl_malloc_device

        integer(qmckl_exit_code) function qmckl_free_device(context, ptr) &
            bind(C, name="qmckl_free_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), value :: context
            type(c_ptr), value :: ptr
        end function qmckl_free_device

        integer(qmckl_exit_code) function qmckl_memcpy_H2D(context, dest, src, size) &
            bind(C, name="qmckl_memcpy_H2D")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), value :: context
            type(c_ptr), value :: dest
            type(c_ptr), value :: src
            integer(c_size_t), value :: size
        end function qmckl_memcpy_H2D

        integer(qmckl_exit_code) function qmckl_memcpy_D2H(context, dest, src, size) &
            bind(C, name="qmckl_memcpy_D2H")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), value :: context
            type(c_ptr), value :: dest
            type(c_ptr), value :: src
            integer(c_size_t), value :: size
        end function qmckl_memcpy_D2H

        integer(qmckl_exit_code) function qmckl_memcpy_D2D(context, dest, src, size) &
            bind(C, name="qmckl_memcpy_D2D")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), value :: context
            type(c_ptr), value :: dest
            type(c_ptr), value :: src
            integer(c_size_t), value :: size
        end function qmckl_memcpy_D2D

    !!!!!!!!!!!
        ! TREXIO
    !!!!!!!!!!!

        integer(qmckl_exit_code) function qmckl_trexio_read_device &
            (context, file_name, size_max) &
            bind(C, name="qmckl_trexio_read_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            integer(c_int64_t), intent(in), value :: size_max
            character(c_char), intent(in)          :: file_name(size_max)
        end function qmckl_trexio_read_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_ao_num_device(context, num) &
            bind(C, name="qmckl_set_ao_basis_ao_num_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            integer(c_int64_t), intent(in), value :: num
        end function qmckl_set_ao_basis_ao_num_device

        integer(qmckl_exit_code) function qmckl_set_electron_coord_device(context, transp, walk_num, coord, size_max) &
            bind(C, name="qmckl_set_electron_coord_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            character, intent(in), value :: transp
            integer(c_int64_t), intent(in), value :: walk_num
            real(c_double), intent(in)          :: coord(*)
            integer(c_int64_t), intent(in), value :: size_max
        end function qmckl_set_electron_coord_device

        integer(qmckl_exit_code) function qmckl_set_point_device(context, transp, num, coord, size_max) &
            bind(C, name="qmckl_set_point_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            character(c_char), intent(in), value :: transp
            integer(c_int64_t), intent(in), value :: num
            real(c_double), intent(in)          :: coord(*)
            integer(c_int64_t), intent(in), value :: size_max
        end function qmckl_set_point_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_type_device(context, type) &
            bind(C, name="qmckl_set_ao_basis_shell_num_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            character(c_signed_char), intent(in), value :: type
        end function qmckl_set_ao_basis_type_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_shell_num_device(context, shell_num) &
            bind(C, name="qmckl_set_ao_basis_shell_num_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            integer(c_int64_t), intent(in), value :: shell_num
        end function qmckl_set_ao_basis_shell_num_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_prim_num_device(context, prim_num) &
            bind(C, name="qmckl_set_ao_basis_prim_num_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            integer(c_int64_t), intent(in), value :: prim_num
        end function qmckl_set_ao_basis_prim_num_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_nucleus_index_device(context, nucleus_index, size_max) &
            bind(C, name="qmckl_set_ao_basis_nucleus_index_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            type(c_ptr), intent(in), value :: nucleus_index ! Elements of type int64_t
            integer(c_int64_t), intent(in), value :: size_max
        end function qmckl_set_ao_basis_nucleus_index_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_nucleus_shell_num_device(context, nucleus_shell_num, nucl_num) &
            bind(C, name="qmckl_set_ao_basis_nucleus_shell_num_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            type(c_ptr), intent(in), value :: nucleus_shell_num ! Elements of type int64_t
            integer(c_int64_t), intent(in), value :: nucl_num
        end function qmckl_set_ao_basis_nucleus_shell_num_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_shell_ang_mom_device(context, shell_ang_mom, shell_num) &
            bind(C, name="qmckl_set_ao_basis_shell_ang_mom_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            type(c_ptr), intent(in), value :: shell_ang_mom ! Elements of type int32_t
            integer(c_int64_t), intent(in), value :: shell_num
        end function qmckl_set_ao_basis_shell_ang_mom_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_shell_factor_device(context, shell_factor, shell_num) &
            bind(C, name="qmckl_set_ao_basis_shell_factor_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            type(c_ptr), intent(in), value :: shell_factor ! Elements of type double
            integer(c_int64_t), intent(in), value :: shell_num
        end function qmckl_set_ao_basis_shell_factor_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_shell_prim_num_device(context, shell_prim_num, shell_num) &
            bind(C, name="qmckl_set_ao_basis_shell_prim_num_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            type(c_ptr), intent(in), value :: shell_prim_num ! Elements of type int64_t
            integer(c_int64_t), intent(in), value :: shell_num
        end function qmckl_set_ao_basis_shell_prim_num_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_shell_prim_index_device(context, shell_prim_index, shell_num) &
            bind(C, name="qmckl_set_ao_basis_shell_prim_index_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            type(c_ptr), intent(in), value :: shell_prim_index ! Elements of type int64_t
            integer(c_int64_t), intent(in), value :: shell_num
        end function qmckl_set_ao_basis_shell_prim_index_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_exponent_device(context, exponent, prim_num) &
            bind(C, name="qmckl_set_ao_basis_exponent_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            type(c_ptr), intent(in), value :: exponent ! Elements of type double
            integer(c_int64_t), intent(in), value :: prim_num
        end function qmckl_set_ao_basis_exponent_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_coefficient_device(context, coefficient, prim_num) &
            bind(C, name="qmckl_set_ao_basis_coefficient_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            type(c_ptr), intent(in), value :: coefficient ! Elements of type double
            integer(c_int64_t), intent(in), value :: prim_num
        end function qmckl_set_ao_basis_coefficient_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_prim_factor_device(context, prim_factor, prim_num) &
            bind(C, name="qmckl_set_ao_basis_prim_factor_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            type(c_ptr), intent(in), value :: prim_factor ! Elements of type double
            integer(c_int64_t), intent(in), value :: prim_num
        end function qmckl_set_ao_basis_prim_factor_device

        integer(qmckl_exit_code) function qmckl_set_ao_basis_ao_factor_device(context, ao_factor, ao_num) &
            bind(C, name="qmckl_set_ao_basis_ao_factor_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            type(c_ptr), intent(in), value :: ao_factor ! Elements of type double
            integer(c_int64_t), intent(in), value :: ao_num
        end function qmckl_set_ao_basis_ao_factor_device

    !!!!!!!!!!!
        ! AO
    !!!!!!!!!!!

        integer(qmckl_exit_code) function qmckl_get_ao_basis_ao_vgl_device(context, ao_vgl, size_max) &
            bind(C, name="qmckl_get_ao_basis_ao_vgl_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            real(c_double), intent(out)         :: ao_vgl(*)
            integer(c_int64_t), intent(in), value :: size_max
        end function qmckl_get_ao_basis_ao_vgl_device

        integer(qmckl_exit_code) function qmckl_get_ao_basis_ao_value_device(context, ao_value, size_max) &
            bind(C, name="qmckl_get_ao_basis_ao_value_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            real(c_double), intent(out)         :: ao_value(*)
            integer(c_int64_t), intent(in), value :: size_max
        end function qmckl_get_ao_basis_ao_value_device

    !!!!!!!!!!!
        ! MO
    !!!!!!!!!!!

        integer(qmckl_exit_code) function qmckl_get_mo_basis_mo_vgl_device(context, mo_vgl, size_max) &
            bind(C, name="qmckl_get_mo_basis_mo_vgl_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            real(c_double), intent(out)         :: mo_vgl(*)
            integer(c_int64_t), intent(in), value :: size_max
        end function qmckl_get_mo_basis_mo_vgl_device

        integer(qmckl_exit_code) function qmckl_get_mo_basis_mo_value_device(context, mo_value, size_max) &
            bind(C, name="qmckl_get_mo_basis_mo_value_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            real(c_double), intent(out)         :: mo_value(*)
            integer(c_int64_t), intent(in), value :: size_max
        end function qmckl_get_mo_basis_mo_value_device

        integer(qmckl_exit_code) function qmckl_get_mo_basis_mo_vgl_inplace_device(context, mo_vgl, size_max) &
            bind(C, name="qmckl_get_mo_basis_mo_vgl_inplace_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            real(c_double), intent(out)         :: mo_vgl(*)
            integer(c_int64_t), intent(in), value :: size_max
        end function qmckl_get_mo_basis_mo_vgl_inplace_device

        integer(qmckl_exit_code) function qmckl_get_mo_basis_mo_value_inplace_device(context, mo_value, size_max) &
            bind(C, name="qmckl_get_mo_basis_mo_value_inplace_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            real(c_double), intent(out)         :: mo_value(*)
            integer(c_int64_t), intent(in), value :: size_max
        end function qmckl_get_mo_basis_mo_value_inplace_device

        integer(qmckl_exit_code) function qmckl_get_mo_basis_mo_num_device(context, mo_num) &
            bind(C, name="qmckl_get_mo_basis_mo_num_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            integer(c_int64_t), intent(in), value :: mo_num
        end function qmckl_get_mo_basis_mo_num_device

        integer(qmckl_exit_code) function qmckl_mo_basis_select_mo_device(context, keep, size_max) &
            bind(C, name="qmckl_get_mo_basis_mo_num_device")
            use, intrinsic :: iso_c_binding
            import
            implicit none

            integer(qmckl_context_device), intent(in), value :: context
            integer(c_int32_t), intent(out)         :: keep(*)
            integer(c_int64_t), intent(in), value :: size_max
        end function qmckl_mo_basis_select_mo_device


    !!!!!!!!!!!
        ! ELECTRON
    !!!!!!!!!!!

    end interface
end module qmckl_gpu

File: src/qmckl_gpu_f.c
// This files contains the C wrappers to be interfaced with Fortran

#include "qmckl_gpu.h"

//**********
// CONTEXT
//**********

qmckl_context_device
qmckl_context_touch_device_f(qmckl_context_device context) {
	return qmckl_context_touch_device(context);
}

qmckl_context_device qmckl_context_create_device_f(int *device_id) {
	return qmckl_context_create_device(*device_id);
}

qmckl_exit_code qmckl_context_destroy_device_f(qmckl_context_device context) {
	return qmckl_context_destroy_device(context);
}

//**********
// MEMORY
//**********

/* Allocs & frees */

void *qmckl_malloc_host_f(qmckl_context_device context,
						  const qmckl_memory_info_struct info);

qmckl_exit_code qmckl_free_host_f(qmckl_context_device context,
								  void *const ptr);

void *qmckl_malloc_device_f(qmckl_context_device context, size_t *size) {
	return qmckl_malloc_device(context, *size);
}

qmckl_exit_code qmckl_free_device_f(qmckl_context_device context,
									void *const ptr) {
	return qmckl_free_device(context, ptr);
}

/* Memcpys */

qmckl_exit_code qmckl_memcpy_H2D_f(qmckl_context_device context,
								   void *const dest, void *const src,
								   size_t *size) {
	return qmckl_memcpy_H2D(context, dest, src, *size);
}

qmckl_exit_code qmckl_memcpy_D2H_f(qmckl_context_device context,
								   void *const dest, void *const src,
								   size_t *size) {
	return qmckl_memcpy_D2H(context, dest, src, *size);
}

qmckl_exit_code qmckl_memcpy_D2D_f(qmckl_context_device context,
								   void *const dest, void *const src,
								   size_t *size) {
	return qmckl_memcpy_D2D(context, dest, src, *size);
}

//**********
// TREXIO
//**********

qmckl_exit_code qmckl_trexio_read_device_f(qmckl_context_device context,
										   char *file_name, int64_t *size_max) {
	return qmckl_trexio_read_device(context, file_name, *size_max);
}

qmckl_exit_code qmckl_set_electron_num_device_f(qmckl_context_device context,
												int64_t *up_num,
												int64_t *down_num) {
	return qmckl_set_electron_num_device(context, *up_num, *down_num);
}

qmckl_exit_code qmckl_set_electron_coord_device_f(qmckl_context_device context,
												  char *transp,
												  int64_t *walk_num,
												  double *coord,
												  int64_t *size_max) {
	return qmckl_set_electron_coord_device(context, *transp, *walk_num, coord,
										   *size_max);
}

qmckl_exit_code qmckl_set_point_device_f(qmckl_context_device context,
										 char *transp, int64_t *num,
										 double *coord, int64_t *size_max) {
	return qmckl_set_point_device(context, *transp, *num, coord, *size_max);
}

qmckl_exit_code qmckl_get_nucleus_num_device_f(qmckl_context_device context,
											   int64_t *num) {
	return qmckl_get_nucleus_num_device(context, num);
}

qmckl_exit_code qmckl_set_nucleus_num_device_f(qmckl_context_device context,
											   int64_t *num) {
	return qmckl_set_nucleus_num_device(context, *num);
}

qmckl_exit_code qmckl_set_nucleus_charge_device_f(qmckl_context_device context,
												  double *charge,
												  int64_t *size_max) {
	return qmckl_set_nucleus_charge_device(contextn charge, *size_max);
}

qmckl_exit_code qmckl_set_nucleus_coord_device_f(qmckl_context_device context,
												 char *transp, double *coord,
												 int64_t *size_max) {
	return qmckl_set_nucleus_coord_device(context, *transp, coord, *size_max);
}

qmckl_exit_code qmckl_get_ao_basis_ao_num_device_f(qmckl_context_device context,
												   int64_t *ao_num) {
	return qmckl_get_ao_basis_ao_num_device(context, ao_num);
}

qmckl_exit_code qmckl_set_ao_basis_type_device_f(qmckl_context_device context,
												 char *basis_type) {
	return qmckl_set_ao_basis_type_device(context, *basis_type);
}

qmckl_exit_code
qmckl_set_ao_basis_shell_num_device_f(qmckl_context_device context,
									  int64_t *shell_num) {
	return qmckl_set_ao_basis_shell_num_device(context, *shell_num);
}

qmckl_exit_code
qmckl_set_ao_basis_prim_num_device_f(qmckl_context_device context,
									 int64_t *prim_num) {
	return qmckl_set_ao_basis_prim_num_device(context, *prim_num);
}

qmckl_exit_code qmckl_set_ao_basis_coefficient_device_f(
	qmckl_context_device context, double *coefficient, int64_t *size_max) {
	return qmckl_set_ao_basis_coefficient_device(context, coefficient,
												 *size_max);
}

qmckl_exit_code qmckl_set_ao_basis_ao_num_device_f(qmckl_context_device context,
												   int64_t *ao_num) {
	return qmckl_set_ao_basis_ao_num_device(context, *ao_num);
}

qmckl_exit_code qmckl_set_ao_basis_nucleus_index_device_f(
	qmckl_context_device context, int64_t *nucleus_index, int64_t *size_max) {
	return qmckl_set_ao_basis_nucleus_index_device(context, nucleus_index,
												   *size_max);
}

qmckl_exit_code
qmckl_set_ao_basis_nucleus_shell_num_device_f(qmckl_context_device context,
											  int64_t *nucleus_shell_num,
											  int64_t *size_max) {
	return qmckl_set_ao_basis_nucleus_shell_num_device(
		context, nucleus_shell_num, *size_max);
}

qmckl_exit_code qmckl_set_ao_basis_shell_ang_mom_device_f(
	qmckl_context_device context, int32_t *shell_ang_mom, int64_t *size_max) {
	return qmckl_set_ao_basis_shell_ang_mom_device(context, shell_ang_mom,
												   *size_max);
}

qmckl_exit_code qmckl_set_ao_basis_shell_prim_num_device_f(
	qmckl_context_device context, int64_t *shell_prim_num, int64_t *size_max) {
	return qmckl_set_ao_basis_shell_prim_num_device(context, shell_prim_num,
													*size_max);
}

qmckl_exit_code qmckl_set_ao_basis_shell_prim_index_device_f(
	qmckl_context context, int64_t *shell_prim_index, int64_t *size_max) {
	qmckl_set_ao_basis_shell_prim_index_device(context, shell_prim_index,
											   *size_max);
}

qmckl_exit_code qmckl_set_ao_basis_shell_factor_device_f(
	qmckl_context_device context, double *shell_factor, int64_t *size_max) {
	return qmckl_set_ao_basis_shell_factor_device(context, shell_factor,
												  *size_max);
}

qmckl_exit_code
qmckl_set_ao_basis_exponent_device_f(qmckl_context_device context,
									 double *exponent, int64_t *size_max) {
	return qmckl_set_ao_basis_exponent_device(context, exponent, *size_max);
}

qmckl_exit_code qmckl_set_ao_basis_prim_factor_device_f(qmckl_context context,
														double *prim_factor,
														int64_t *size_max) {
	return qmckl_set_ao_basis_prim_factor_device(context, prim_factor,
												 *size_max);
}

qmckl_exit_code
qmckl_set_ao_basis_ao_factor_device_f(qmckl_context_device context,
									  double *ao_factor, int64_t *size_max) {
	return qmckl_set_ao_basis_ao_factor_device(context, ao_factor, *size_max);
}

qmckl_exit_code
qmckl_get_ao_basis_shell_num_device_f(qmckl_context_device context,
									  int64_t *ao_num) {
	return qmckl_get_ao_basis_shell_num_device(context, ao_num);
}

qmckl_exit_code
qmckl_get_ao_basis_prim_num_device_f(qmckl_context_device context,
									 int64_t *prim_num) {
	return qmckl_get_ao_basis_prim_num_device(context, prim_num);
}

qmckl_exit_code qmckl_get_ao_basis_type_device_f(qmckl_context_device context,
												 char *type) {
	return qmckl_get_ao_basis_type_device(context, type);
}

qmckl_exit_code
qmckl_get_ao_basis_nucleus_shell_num_device_f(qmckl_context_device context,
											  int64_t *nucleus_shell_num,
											  int64_t *nucl_num) {
	return qmckl_get_ao_basis_nucleus_shell_num_device(
		context, nucleus_shell_num, *nucl_num);
}

qmckl_exit_code qmckl_get_ao_basis_shell_ang_mom_device_f(
	qmckl_context_device context, int32_t *shell_ang_mom, int64_t *shell_num) {
	return qmckl_get_ao_basis_shell_ang_mom_device(context, shell_ang_mom,
												   *shell_num);
}

qmckl_exit_code qmckl_get_ao_basis_shell_factor_device_f(
	qmckl_context_device context, double *shell_factor, int64_t *shell_num) {
	return qmckl_get_ao_basis_shell_factor_device(context, shell_factor,
												  *shell_num);
}

qmckl_exit_code qmckl_get_ao_basis_shell_prim_num_device_f(
	qmckl_context_device context, int64_t *shell_prim_num, int64_t *shell_num) {
	return qmckl_get_ao_basis_shell_prim_num_device(context, shell_prim_num,
													*shell_num);
}

qmckl_exit_code
qmckl_get_ao_basis_shell_prim_index_device_f(qmckl_context_device context,
											 int64_t *nucleus_shell_num,
											 int64_t *shell_num) {
	return qmckl_get_ao_basis_shell_prim_index_device(
		context, nucleus_shell_num, *shell_num);
}

qmckl_exit_code
qmckl_get_ao_basis_exponent_device_f(qmckl_context_device context,
									 double *exponent, int64_t *prim_num) {
	return qmckl_get_ao_basis_exponent_device(context, exponent, *prim_num);
}

qmckl_exit_code qmckl_get_ao_basis_coefficient_device_f(
	qmckl_context_device context, double *coefficient, int64_t *prim_num) {
	return qmckl_get_ao_basis_coefficient_device(context, coefficient,
												 *prim_num);
}

qmckl_exit_code qmckl_get_ao_basis_prim_factor_device_f(
	qmckl_context_device context, double *prim_factor, int64_t *prim_num) {
	return qmckl_get_ao_basis_prim_factor_device(context, prim_factor,
												 *prim_num);
}

qmckl_exit_code
qmckl_get_ao_basis_ao_factor_device_f(qmckl_context_device context,
									  double *ao_factor, int64_t *ao_num) {
	return qmckl_get_ao_basis_ao_factor_device(context, ao_factor, *ao_num);
}

qmckl_exit_code qmckl_get_ao_basis_nucleus_index_device_f(
	qmckl_context_device context, int64_t *nucleus_index, int64_t *nucl_num) {
	return qmckl_get_ao_basis_nucleus_index_device(context, nucleus_index,
												   *nucl_num);
}

qmckl_exit_code qmckl_set_mo_basis_coefficient_device_f(qmckl_context context,
														double *coefficient) {
	return qmckl_set_mo_basis_coefficient_device(context, coefficient);
}

//**********
// AO
//**********

qmckl_exit_code qmckl_get_ao_basis_ao_num_device_f(qmckl_context_device context,
												   int64_t *ao_num) {
	return qmckl_get_ao_basis_ao_num_device(context, ao_num);
}

qmckl_exit_code qmckl_get_ao_basis_ao_vgl_device_f(qmckl_context_device context,
												   double *ao_vgl,
												   int64_t *size_max) {
	return qmckl_get_ao_basis_ao_vgl_device(context, ao_vgl, *size_max);
}

qmckl_exit_code
qmckl_get_ao_basis_ao_value_device_f(qmckl_context_device context,
									 double *ao_vgl, int64_t *size_max) {
	return qmckl_get_ao_basis_ao_value_device(context, ao_vgl, *size_max);
}

//**********
// MO
//**********

qmckl_exit_code qmckl_get_mo_basis_mo_vgl_device_f(qmckl_context context,
												   double *const mo_vgl,
												   const int64_t *size_max) {
	return qmckl_get_mo_basis_mo_vgl_device(context, mo_vgl, *size_max);
}

qmckl_exit_code
qmckl_get_mo_basis_mo_value_device_f(qmckl_context_device context,
									 double *mo_value, int64_t *size_max) {
	return qmckl_get_mo_basis_mo_value_device(context, mo_vgl, *size_max);
}

qmckl_exit_code qmckl_get_mo_basis_mo_value_inplace_device_f(
	qmckl_context_device context, double *mo_value, int64_t *size_max) {
	return qmckl_get_mo_basis_mo_value_inplace_device(context, mo_value,
													  *size_max);
}

qmckl_exit_code qmckl_get_mo_basis_mo_vgl_inplace_device_f(
	qmckl_context context, double *const mo_vgl, const int64_t *size_max) {
	return qmckl_get_mo_basis_mo_vgl_inplace_device(context, mo_vgl, *size_max);
}
